### PR TITLE
Add Elixir solution with generalized iteration

### PR DESCRIPTION
**Pull request by jonatanklosko**

### CODE REVIEW

File: elixir/put_cons.ex [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+      %{"name" => "Immutability", "position" => 2}
+    ]
+  }
+]
+
+defmodule DataTraversal do
+  def update_positions(sections) do
+    sections
+    |> put_cons(["position"], 1)
+    |> chunk_when(& &1["reset_lesson_position"])
```
Comment (jonatanklosko) [inline elixir/put_cons.ex:62]: Initially I thought I could do it with `Enum.chunk_by`, but then I realised it's not appropriate here, so this turned out quite overengineered :see_no_evil:

Comment (nickjj) [reply elixir/put_cons.ex:62]: There's also currently an `Enum.chunk_by` solution in the repo already btw.

**Comment by nickjj** [comment]
Hi, thanks for submitting a solution.

This one is a bit above my paygrade but what do you think about renaming the file from `put_cons.ex` to `chunk_while_map_reduce.ex`? My thought process there is, these files should be named based on the main function(s) you used to arrive at your solution. If you have a better suggestion by all means go for it.

File: elixir/put_cons.ex (hunk 1 above)
Comment (jonatanklosko) [reply elixir/put_cons.ex:62]: I see, that unfortunately breaks if the resetting section goes in the middle.

**Comment by jonatanklosko** [comment]
Wasn't sure about the filename either, because the `put_cons` function is more about pattern matching and composition, than `map_reduce`, but let's go with `chunk_while_map_reduce` :>

**Comment by nickjj** [comment]
Ah ok, that's a good point about pattern matching and after looking at the code I fully agree. Would `chunk_while_pattern_matching.ex` be more appropriate?

**Comment by jonatanklosko** [comment]
Sounds better! Ideally it would be `chunk_while_nested_increment`, but that's more specific to the problem rather than the solution ^^

**Comment by nickjj** [comment]
Thanks!

**Pull request merged**